### DISCUSSION TITLE
Question about BRMS API authentication with PAT

### DISCUSSION THREAD

**Discussion by mm13-github** [gorules/zen#147] [Q&A] [answered]
I'm doing a feasibility study for the adoption of GoRules within a project, and I'm pleasantly surprised by its features and performance. 

I'm encountering an issue with using the APIs, specifically regarding authentication. I understand that the mechanism varies depending on whether one is using APIs related to RELEASE compared to the EVALUATE API. In the latter case, I am able to authenticate successfully with the project token. However, in the first case, I am using the PAT (Personal Access Token) as indicated in the documentation:

https://gorules.io/docs/brms/gorules-brms-api#personal-access-token-pat

 but I always receive the following error:
 
 curl --request GET \
  --url http://localhost:9080/api/projects/0d42b18c-2b84-40bb-84c4-1b9f2eb1cf86/releases \
  --header 'Accept: application/json' \
  --header 'Authorization: Bearer grl_pat__XXXXX'
 
 {"level":40,"time":1712878556989,"pid":1,"hostname":"526de42e67d9","reqId":"9fa50e44-cce8-4350-a26a-94eeceb9f96a","method":"GET","path":"/api/projects/:id/releases","err":{"type":"Error","message":"unauthorized","stack":"Error: unauthorized\n    at /app/main.js:18:44269\n    at step (/app/main.js:18:42447)\n    at Object.next (/app/main.js:18:41749)\n    at auth_middleware_asyncGeneratorStep (/app/main.js:18:41043)\n    at _next (/app/main.js:18:41282)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","data":null,"isBoom":true,"isServer":false,"output":{"statusCode":401,"payload":{"statusCode":401,"error":"Unauthorized","message":"unauthorized"},"headers":{}}},"msg":"request failed"}
 
 What am I doing wrong?
 Thanks

**Comment by ivanmiletic** [comment]
Hi @mm13-github thanks for the feedback,

Curl is from the docs https://gorules.io/docs/brms/list-releases (you can find it top right):

```
curl -L -X GET 'http://localhost:9080/api/projects/0d42b18c-2b84-40bb-84c4-1b9f2eb1cf86/releases' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer <TOKEN>'
```
I have tested with your command on my machine (local) and our Azure cloud and it works.

This error can happen if:
- Your token is invalid

Can you please generate new token and try again, make sure you are pasting properly and also make sure that it has access to all your projects when generating it.

**Reply by mm13-github** [reply]
HI @ivanmiletic,

thanks for the help you are giving me. As suggested I created a token again:

![image](https://github.com/gorules/zen/assets/62216807/c2cbeaef-23c3-4cf6-836e-8f161e791fb7)

![image](https://github.com/gorules/zen/assets/62216807/ce7fd089-99b0-4297-97c3-7ffabb4693d0)

I'm verifed also in database and all data seems correct:

PROJECT:

![image](https://github.com/gorules/zen/assets/62216807/f1cc492b-c45d-4cc8-90b9-11947b95ca5c)

PERSONAL_ACCESS_TOKEN:

![image](https://github.com/gorules/zen/assets/62216807/0e0c0092-102e-4cd8-9314-65e4340859dc)

Unfortunately it continued to have the same error:

```shell
curl -L -X GET 'http://localhost:9080/api/projects/0d42b18c-2b84-40bb-84c4-1b9f2eb1cf86/releases' \
-H 'Accept: application/json' \
-H 'Authorization: Bearer grl_pat_iXw .......... rWxyA'
``` 

```
{"level":40,"time":1712915971773,"pid":1,"hostname":"526de42e67d9","reqId":"12930e99-c7df-49d1-8ffa-578b9ab6b973","method":"GET","path":"/api/projects/:id/releases","err":{"type":"Error","message":"unauthorized","stack":"Error: unauthorized\n    at /app/main.js:18:44269\n    at step (/app/main.js:18:42447)\n    at Object.next (/app/main.js:18:41749)\n    at auth_middleware_asyncGeneratorStep (/app/main.js:18:41043)\n    at _next (/app/main.js:18:41282)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)","data":null,"isBoom":true,"isServer":false,"output":{"statusCode":401,"payload":{"statusCode":401,"error":"Unauthorized","message":"unauthorized"},"headers":{}}},"msg":"request failed"}
``` 

Thanks again,
Massimo

---

Other details:
DOCKER IMAGE = gorules/brms:latest
NGINX_VERSION = 1.24.0
NJS_VERSION =  0.7.12

**Reply by ivanmiletic** [reply, marked as answer]
Hi @mm13-github ,

Strange, i'm not able to reproduce on our systems. Can you send email to support@gorules.io, we can continue from there. Lets organise a 30 minute call next week to troubleshoot your issue.

**Reply by mm13-github** [reply]
Hi @ivanmiletic, 
thank you for your attention and the offer of support, it's very kind of you and I really appreciate it. 

In the next few days, we will contact support@gorules.io and we are still interested in subscribing to a TEAM plan to start testing the multi-environment feature. 

Thank you very much.

**Reply by LeahHeadd** [reply]
was there ever a resolution to this? I'm encountering the same issue

<img width="561" alt="image" src="https://github.com/user-attachments/assets/ab19d326-ea59-4013-9738-daaf1348d0d0">


calling
```
curl -L -X GET "https://${DOMAIN}.gorules.io/api/projects/${PROJECT_ID}/releases" \
> -H "Accept: application/json" \
> -H "Authorization: Bearer ${GRL_PAT}"
```

yields
`{"message":"unauthorized","data":null}`

**Reply by stefan-gorules** [reply]
Hi, it would be indeed good to jump on the call to try and diagnose the issue as we are having trouble replicating it. It could be possible that the underlying user on which token is generated doesn't have permission to access project / releases on that project.

@LeahHeadd tried generating a PAT and fetching releases, however it succeeded. Any other specifics that you can share (e.g. owner of token is has x role on project and y role on GoRules environment).
